src/astarte_message_hub.rs

```diff
@@ -0,0 +1,226 @@
+/*
+ * This file is part of Astarte.
+ *
+ * Copyright 2022 SECO Mind Srl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use log::info;
+use tokio::sync::RwLock;
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::{Request, Response, Status};
+
+use crate::data::astarte::{AstartePublisher, AstarteSubscriber};
+use uuid::Uuid;
+
+use crate::proto_message_hub::message_hub_server::MessageHub;
+use crate::proto_message_hub::{AstarteMessage, Interface, Node};
+
+pub struct AstarteMessageHub<T: AstartePublisher + AstarteSubscriber> {
+    pub nodes: Arc<RwLock<HashMap<Uuid, AstarteNode>>>,
+    pub astarte_handler: T,
+}
+
+pub struct AstarteNode {
+    pub id: Uuid,
+    pub introspection: Vec<Interface>,
+}
+
+impl<T> AstarteMessageHub<T>
+where
+    T: AstartePublisher + AstarteSubscriber,
+{
+    fn new(astarte_handler: T) -> Self {
+        AstarteMessageHub {
+            nodes: Arc::new(RwLock::new(HashMap::new())),
+            astarte_handler,
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl<T: AstartePublisher + AstarteSubscriber + 'static> MessageHub for AstarteMessageHub<T> {
+    type AttachStream = ReceiverStream<Result<AstarteMessage, Status>>;
+
+    async fn attach(&self, request: Request<Node>) -> Result<Response<Self::AttachStream>, Status> {
+        info!("Node Attach Request => {:?}", request);
+        let node = request.into_inner();
+
+        let id = Uuid::parse_str(&node.uuid).map_err(|err| {
+            Status::invalid_argument(format!(
+                "Unable to parse UUID value, err {:?}",
+                err.to_string()
+            ))
+        })?;
+
+        let astarte_node = AstarteNode {
+            id,
+            introspection: node.introspection,
+        };
+
+        let subscribe_result = self.astarte_handler.subscribe(&astarte_node).await;
+
+        if let Ok(rx) = subscribe_result {
+            let mut nodes = self.nodes.write().await;
+            nodes.insert(astarte_node.id.to_owned(), astarte_node);
+            Ok(Response::new(ReceiverStream::new(rx)))
+        } else {
+            Err(Status::internal(format!(
```

Comment:
An error in `subscribe_result` may also come from a reason not due to the message hub (e.g. invalid UUID --> `invalid_argument()`), returning an error status different than `internal` would be more informative